tests/gtest/avifalphamulttest.cc

```diff
@@ -0,0 +1,66 @@
+// Copyright 2022 Google LLC
+// SPDX-License-Identifier: BSD-2-Clause
+
+#include "avif/avif.h"
+#include "aviftest_helpers.h"
+#include "gtest/gtest.h"
+
+namespace libavif {
+namespace {
+
+//------------------------------------------------------------------------------
+
+TEST(AlphaMultiplyTest, OpaqueIsNoOp) {
+  for (bool premultiplied_input : {false, true}) {
+    if (!premultiplied_input) {
+      // TODO(yguyon): Fix the issue and remove this condition.
+      //               The issue is that avifPrepareReformatState() will result
+      //               in different YUV(A)-to-RGB conversion algorithm choices
+      //               (built-in or libyuv) depending on the presence of an
+      //               alpha channel, no matter if it is opaque or not, because
+      //               it considers that unpremultiplied YUVA samples should be
+      //               premultiplied before discarding alpha and converting to
+      //               RGB.
+      continue;
+    }
+
+    // YUVA.
+    testutil::AvifImagePtr opaque_alpha = testutil::CreateImage(
+        1024, 1024, /*depth=*/8, AVIF_PIXEL_FORMAT_YUV444, AVIF_PLANES_ALL);
+    const uint32_t yuva[] = {255, 255, 255, 255};
+    testutil::FillImagePlain(opaque_alpha.get(), yuva);
+    opaque_alpha->alphaPremultiplied = premultiplied_input;
+
+    // View on YUV (no alpha), to make sure the color values are identical.
+    testutil::AvifImagePtr no_alpha(avifImageCreateEmpty(), avifImageDestroy);
+    const avifCropRect rect = {0, 0, opaque_alpha->width, opaque_alpha->height};
+    ASSERT_EQ(avifImageSetViewRect(no_alpha.get(), opaque_alpha.get(), &rect),
+              AVIF_RESULT_OK);
+    avifImageFreePlanes(no_alpha.get(), AVIF_PLANES_A);
+
+    // Decorrelate YUV and Alpha.
+    testutil::FillImageGradient(no_alpha.get());
+
+    // Convert to RGB and discard any Alpha.
+    testutil::AvifRgbImage opaque_rgb(opaque_alpha.get(), opaque_alpha->depth,
+                                      AVIF_RGB_FORMAT_RGB);
+    ASSERT_EQ(avifImageYUVToRGB(opaque_alpha.get(), &opaque_rgb),
+              AVIF_RESULT_OK);
+    testutil::AvifRgbImage no_alpha_rgb(no_alpha.get(), no_alpha->depth,
+                                        AVIF_RGB_FORMAT_RGB);
+    ASSERT_EQ(avifImageYUVToRGB(no_alpha.get(), &no_alpha_rgb), AVIF_RESULT_OK);
+
+    // YUV samples with opaque alpha and with missing alpha should be converted
+    // to the same RGB values.
```

Comment:
This goal is reasonable, but it is also reasonable to expect that a missing alpha channel and an opaque alpha channel may be considered as different in some cases. We don't need to overly constrain our implementation if this goal becomes inconvenient.

It may be better to tolerate small differences so that we can use different conversion functions. The `avifImageIsOpaque()` check is not free. If an opaque alpha plane is uncommon, then it may not make sense to perform this check.

Reply:
> it is also reasonable to expect that a missing alpha channel and an opaque alpha channel may be considered as different in some cases.

Do you have examples of such cases?

> We don't need to overly constrain our implementation if this goal becomes inconvenient.

I was very surprised when I got different results (visible difference with a simple gradient exported to a local png). I thought there was a bug and I spent time figuring out if I did something wrong or not. I believe it would be better to fix this to avoid other people spending time on this topic too in the future. Maybe there could be a visible-enough disclaimer instead of a fix, but at the same time, using user attention for that seems unreasonable.

> It may be better to tolerate small differences so that we can use different conversion functions.

We could perform this test by disabling any use of libyuv and libsharpyuv, but I guess it may still use one of the "fast" built-in conversion functions instead of the slow generic 

> The `avifImageIsOpaque()` check is not free. If an opaque alpha plane is uncommon, then it may not make sense to perform this check.

`avifImageIsOpaque()` is kind of free for translucent images: the odds of having a nearly opaque image with translucent pixels only at the end of the buffer are fairly small. Otherwise it will return false very quickly.

For opaque images it will check all the pixels, but as you said it might be uncommon. And the benefit would be to use libyuv in that case instead of the slow built-in conversion, so probably faster overall.